### ISSUE TITLE
Make all SendAsync methods virtual

### ISSUE THREAD

**Issue by patelmitul** [dotnet/runtime#58647]
I am requesting to make all SendAsync methods virtual in System.Net.Http.HttpClient . Only one of the 4 SendAsync method is virtual. If not all then make the the following SendAsync virtual.

public **_virtual_** Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)

**Comment by danmoseley** [comment]
@patelmitul thanks for the contribution, but we don't make API change decisions in PR's. We have to reach consensus in an issue, and go through API review.
Please see here: https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md

**Comment by karelz** [comment]
@danmoseley this is not PR, but an issue. Reopening ...

**Comment by karelz** [comment]
There is only 1 method virtual, as it is the most general one needed. Note that it overrides base class virtual method `HttpMessageInvoker.SendAsync`:
https://github.com/dotnet/runtime/blob/13d5fa0938506d55a41a32384af0c44ad3740054/src/libraries/System.Net.Http/ref/System.Net.Http.cs#L218
All the other overloads eventually call into it, so there is no need to make the other methods virtual.

Why do you think you need them to be virtual?
If you want to derive and override behavior, all you need to do is override this one and all else should work out.

**Comment by danmoseley** [comment]
Oops 😊  Apologies @patelmitul.

**Comment by patelmitul** [comment]
> If you want to derive and override behavior, all you need to do is override this one and all else should work out.

It would not work if I override it in derived class. If I call GetAsync or GetStreamAsync of HttpClient then it will not call the derived SendAsync method of sub class because base HttpClient class has not allowed it. Overriding without base method being virtual is not true polymorphism.

Making it virtual will allow little more flexibility on per project need.

**Comment by karelz** [comment]
@patelmitul all the SendAsync overloads call the one which is virtual on base class. Check the code, or try it out.
Same holds for `Send`. The important one is virtual on base class.

**Comment by patelmitul** [comment]
GetAsync calls the SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken).
That SendAsync will call the base SendAsync which is virtual. But, If I override the only SendAsync of the base class then it will not be called by the GetAsync. 
The base class's SendAsync is overridden in HttpClient which calls the overload with HttpCompletionOption. So, overriding the only virtual method does not work as one would expect for GetAsync.

**Comment by karelz** [comment]
You are right, `HttpClient` intentionally calls `base.SendAsync` to bypass overrides.
Stepping back a moment, it is not recommended to override this method, but rather leverage `HttpMessageHandler` extension model -- see [HttpClient docs](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-5.0#remarks):

> Derived classes should not override the virtual methods on the class. Instead, use a constructor overload that accepts [HttpMessageHandler](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpmessagehandler?view=net-5.0) to configure any pre- or post-request processing.

Is that what you are going after perhaps?

**Comment by patelmitul** [comment]
I agree about HttpMessageHandler and I use it because it provides more flexibility over overloading SendAsync. 
But, I have seen enough projects where people have tried to derive a class from HttpClient and that prompted me to raise this issue. If this class is not meant to be extended then it needs to be sealed otherwise it needs to provide functionality that one would expect from inheritance.
At this point the title of this topic and discussion are not in sync.

**Comment by karelz** [comment]
@patelmitul I think there may be other reasons people might want to inherit from `HttpClient` -- e.g. adding more verb methods. Though extension methods may be used as well.
I agree that if we designed the class from scratch, we would probably seal it, or we would think about extensibility in different ways. Sadly, we do not have time machine.
Making the class sealed at this point would be breaking change, so not something worth it IMO.

Quick internet search did not bring up too many attempts to inherit from `HttpClient`:
- https://stackoverflow.com/questions/15716399/extend-httpclient (10 upvotes)
- https://www.reddit.com/r/dotnet/comments/lw16vv/how_would_one_inherit_from_httpclient/ (2 upvotes)
- https://stackoverflow.com/questions/68292933/net-services-addhttpclient-automatic-access-token-handling (0 upvotes)

My recommendation is to close the issue as Won't Fix.

**Comment by karelz** [comment]
Triage: We confirmed on the team, that the only reason to inherit from `HttpClient` today is to add more utility methods (something achievable also via static extension methods).
We do not plan to make the methods virtual -- it may be useful for mocking (though one can achieve that via Handler extensibility as well). There is related existing ask for interface on `HttpClient` which should have this ask added as +1.
Closing.
